src/connector/src/sink/mongodb.rs

```diff
@@ -0,0 +1,750 @@
+        let schema = param.schema();
+        Ok(Self {
+            config,
+            param,
+            schema,
+            pk_indices,
+            is_append_only,
+        })
+    }
+}
+
+impl TryFrom<SinkParam> for MongodbSink {
+    type Error = SinkError;
+
+    fn try_from(param: SinkParam) -> std::result::Result<Self, Self::Error> {
+        MongodbSink::new(param)
+    }
+}
+
+impl Sink for MongodbSink {
+    type Coordinator = DummySinkCommitCoordinator;
+    type LogSinker = LogSinkerOf<MongodbSinkWriter>;
+
+    const SINK_NAME: &'static str = MONGODB_SINK;
+
+    fn is_sink_decouple(_desc: &SinkDesc, user_specified: &SinkDecouple) -> Result<bool> {
+        match user_specified {
+            // Set default sink decouple to false, because mongodb sink writer only ensure delivery on checkpoint barrier
+            SinkDecouple::Default => Ok(false),
+            SinkDecouple::Disable => Ok(false),
+            SinkDecouple::Enable => Ok(true),
+        }
+    }
+
+    async fn validate(&self) -> Result<()> {
+        if !self.is_append_only {
+            if self.pk_indices.is_empty() {
+                return Err(SinkError::Config(anyhow!(
+                    "Primary key not defined for upsert mongodb sink (please define in `primary_key` field)")));
+            }
+
+            // checking if there is a non-pk field's name is `_id`
+            if self
+                .schema
+                .fields
+                .iter()
+                .enumerate()
+                .any(|(i, field)| !self.pk_indices.contains(&i) && field.name == MONGODB_PK_NAME)
+            {
+                return Err(SinkError::Config(anyhow!(
+                    "_id field must be the sink's primary key, but a non primary key field name is _id",
+                )));
+            }
+
+            // assume the sink's pk is (a, b) and then the data written to mongodb will be
+            // { "_id": {"a": 1, "b": 2}, "a": 1, "b": 2, ... }
+            // you can see that the compound pk (a, b) is turned into an Object {"a": 1, "b": 2}
+            // and the each pk field is become as a field of the document
+            // but if the sink's pk is (_id, b) and the data will be:
+            // { "_id": {"_id": 1, "b": 2}, "b": 2, ... }
+            // in this case, the original _id field of the compound pk has been overridden
+            // we should consider this is a schema error
+            if self.pk_indices.len() > 1
+                && self
+                    .pk_indices
+                    .iter()
+                    .map(|&idx| self.schema.fields[idx].name.as_str())
+                    .any(|field| field == MONGODB_PK_NAME)
+            {
+                return Err(SinkError::Config(anyhow!(
+                    "primary key fields must not contain a field named _id"
+                )));
+            }
+        }
+
+        if self.config.bulk_write_max_entries > MONGODB_BULK_WRITE_SIZE_LIMIT {
```

Comment:
Maybe we can do `client.database(target_db).list_collections(None, None)` to test the reachability to both the cluster and also target database. 

https://docs.rs/mongodb/latest/mongodb/struct.Database.html#method.list_collections

Reply:
> Is this limit a hard limit suggested by any official mongodb document?

actually the limit is 100,000 (you can find the related doc in [hello](https://www.mongodb.com/docs/v4.2/reference/command/hello/#hello.maxWriteBatchSize) command), I have tried sending more than 100,000 records in a command, and you will be notified that document count cannot exceed 100,000. Therefore, I think 65536 should be a reasonable limit. Please note, there is also a bytes limitation called `maxMessageSizeBytes` which is default to 48,000,000.  

> Maybe we can do client.database(target_db).list_collections(None, None) to test the reachability to both the cluster and also target database.

[hello](https://www.mongodb.com/docs/v4.2/reference/command/hello/) is actually a replication command that clients typically use to determine server-side states. It can provide information such as the instance type (mongod or mongos) and whether the connected instance is writable, etc. Although there is also a `ping` command can check the reachability, but I think `hello` is more suitable for this case, because it give us an opportunity to check whether the instance is writable. If the instance is not writable, sink may report a user error immediately.

Reply:
Thanks for the explanation! Could you please add a comment on the hard limit const and paste the link there?